text/0052-reserved-namespaces.md

```diff
@@ -0,0 +1,108 @@
+
+## Detailed design
+
+We will update the [namespace](https://docs.cedarpolicy.com/policies/syntax-grammar.html#grammar-path) to the following to exclude namespaces starting with `__`.
+
+```
+Path ::= LEADING_NAMESPACE_IDENT {'::' IDENT}
+LEADING_NAMESPACE_IDENT ::= IDENT - RESERVED_NAMESPACE_IDENT
+RESERVED_NAMSPACE_IDENT ::= ['_']['_'].*
+IDENT ::= ['_''a'-'z''A'-'Z']['_''a'-'z''A'-'Z''0'-'9']* - RESERVED
+RESERVED ::= 'true' | 'false' | 'if' | 'then' | 'else' | 'in' | 'like' | 'has'
+```
+
+This primarily applies to entity types names.
+In the case where an entity type isn't qualified with a namespace, the leading identifier is just the entity type, so we would forbid an unqualified type `__User`.
+It also applies anywhere else namespaces may appear, including the top-level `namespace` element in a schema file, qualified references to common types in a schema file, and extension function names in policies.
+
+It does not apply where a standalone identifier is expected.
+For example, an entity attribute could still start with `__`.
+
+## Drawbacks
+
+This is a significant breaking change.
+A user with namespaces starting with `__` will need to update their policies, schema, and entity data to use an unreserved namespace.
+The nature of this breaking change fortunately allows for migration away from reserved namespaces before the change is released,
+and the break will manifest immediately on attempting to parse an affected policy, schema, or entity.
+
+## Alternatives
+
+
+### A larger breaking change
+
+We could make a larger breaking change by reserving all _identifiers_ starting with `__` instead of just namespaces.
+This would include inner namespace elements and attribute names on entities and records.
+For example, we would additional forbid referencing an entity `namespace::__User` or writing a record literal `{__attr: 1}`.
+We don't currently see an immediate need for this more aggressive change, but making it now provides even more flexibility in future changes at the cost of the breaking change affecting more users.
+
+### A smaller breaking change
+
+We can make a smaller breaking change by reserving only the `__cedar` namespace.
+Only identifiers whose path starts with `__cedar` would become invalid identifiers, leaving users free to use an identifier like `__namespace::User`, but not `__cedar::User` or `__cedar::inner::User`.
+This still provides names we can use in future features while only affecting users who specifically use this exact namespace.
+
+### No breaking change
+
+We can avoid a breaking change to Cedar by updating the Grammar to allow for a new identifier that is currently invalid, and then reserve that identifier.
+For example, we could reserve `$cedar` or `cedar$` (other options for characters include `%` or `!`).
+It would would be an error to use this identifier in any way that isn't explicitly permitted by the Cedar spec.
```

Comment:
I also think this would be a completely reasonable alternative, if we can settle on a suitable identifier.  (There is some concern that `$cedar` looks like a macro invocation or template parameter in other languages.)

Reply:
This implementation of this alternative is hard to get right as we need to make exceptions on the grammar level.